File: scraper.js
const { GoogleSpreadsheet } = require('google-spreadsheet');
const { JWT } = require('google-auth-library');
const gplay = require('google-play-scraper');

async function main() {
    // 1. 구글 인증 및 시트 연결
    const creds = JSON.parse(process.env.GCP_CREDENTIALS);
    const serviceAccountAuth = new JWT({
        email: creds.client_email,
        key: creds.private_key,
        scopes: ['https://www.googleapis.com/auth/spreadsheets'],
    });

    const sheetUrl = process.env.SHEET_URL;
    const docId = sheetUrl.match(/\/d\/([a-zA-Z0-9-_]+)/)[1];
    const doc = new GoogleSpreadsheet(docId, serviceAccountAuth);
    await doc.loadInfo();

    // 2. 한국 시간(KST) 기준으로 날짜 설정
    const now = new Date(new Date().getTime() + 9 * 60 * 60 * 1000);
    const todayStr = now.toISOString().split('T')[0];
    
    const yesterday = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    const yesterdayStr = yesterday.toISOString().split('T')[0];

    // 3. 어제 시트에서 전일 순위 파악
    const previousRanks = {};
    const yesterdaySheet = doc.sheetsByTitle[yesterdayStr];
    
    if (yesterdaySheet) {
        const rows = await yesterdaySheet.getRows();
        for (const row of rows) {
            const appName = row.get('앱 이름');
            const rank = row.get('순위');
            if (appName && rank) {
                previousRanks[appName] = parseInt(rank, 10);
            }
        }
    } else {
        console.log(`어제(${yesterdayStr}) 시트가 없어 전일 대비 데이터는 'NEW'로 표시됩니다.`);
    }

    // 4. 오늘 시트 준비 (초기화 또는 생성)
    let todaySheet = doc.sheetsByTitle[todayStr];
    const headers = ["날짜", "순위", "앱 이름", "개발사(퍼블리셔)", "전일 대비"];
    
    if (todaySheet) {
        await todaySheet.clear();
        await todaySheet.setHeaderRow(headers);
    } else {
        todaySheet = await doc.addSheet({ title: todayStr, headerValues: headers });
    }

    // 5. 구글 플레이스토어 크롤링 (1~100위)
    console.log("플레이스토어 데이터를 가져오는 중...");
    const topGames = await gplay.list({
        category: 'GAME',
        collection: 'TOPGROSSING',
        num: 100,
        country: 'kr',
        lang: 'ko'
    });

    // 6. 데이터 계산 및 추가
    const newRows = [];
    for (let i = 0; i < topGames.length; i++) {
        const game = topGames[i];
        const rank = i + 1;
        const title = game.title;
        const developer = game.developer || '알 수 없음';
        
        let change = 'NEW';
        if (previousRanks[title]) {
            const diff = previousRanks[title] - rank;
            if (diff > 0) change = `▲ ${diff}`;
            else if (diff < 0) change = `▼ ${Math.abs(diff)}`;
            else change = '-';
        }

        newRows.push({
            "날짜": todayStr,
            "순위": rank,
            "앱 이름": title,
            "개발사(퍼블리셔)": developer,
            "전일 대비": change
        });
    }

    if (newRows.length > 0) {
        await todaySheet.addRows(newRows);
        console.log(`${todayStr} 기준 Top 100 업데이트 완료!`);
    } else {
        console.log("업데이트할 데이터가 없습니다.");
    }
}

main().catch(console.error);

File: scraper.py
import gspread
from gspread.exceptions import WorksheetNotFound
from oauth2client.service_account import ServiceAccountCredentials
from google_play_scraper import collection, Collection
from datetime import datetime, timedelta, timezone
import os
import json

# 1. 인증 및 구글 시트 연결
scope = ["https://spreadsheets.google.com/feeds", "https://www.googleapis.com/auth/drive"]
creds_json = os.environ.get('GCP_CREDENTIALS')
creds_dict = json.loads(creds_json)
creds = ServiceAccountCredentials.from_json_keyfile_dict(creds_dict, scope)
client = gspread.authorize(creds)

sheet_url = os.environ.get('SHEET_URL')
doc = client.open_by_url(sheet_url)

# 2. 한국 시간(KST) 기준으로 오늘/어제 날짜 계산
KST = timezone(timedelta(hours=9))
now = datetime.now(KST)
today_str = now.strftime("%Y-%m-%d")
yesterday_str = (now - timedelta(days=1)).strftime("%Y-%m-%d")

# 3. 어제 시트에서 전일 순위 파악하기 (앱 이름 기준)
previous_ranks = {}
try:
    yesterday_sheet = doc.worksheet(yesterday_str)
    records = yesterday_sheet.get_all_records()
    for row in records:
        if row.get('앱 이름') and row.get('순위'):
            previous_ranks[row['앱 이름']] = int(row['순위'])
except WorksheetNotFound:
    print(f"어제({yesterday_str}) 시트를 찾을 수 없어 전일 대비 데이터는 'NEW'로 표시됩니다.")

# 4. 오늘 날짜의 시트 준비 (없으면 새로 생성, 있으면 초기화)
try:
    today_sheet = doc.worksheet(today_str)
    today_sheet.clear() # 이미 시트가 있다면 내용을 비우고 덮어씌움
except WorksheetNotFound:
    # 100위까지 기록해야 하므로 넉넉하게 120행, 5열로 시트 생성
    today_sheet = doc.add_worksheet(title=today_str, rows="120", cols="5")

# 5. 구글 플레이스토어 게임 최고 매출 순위 크롤링 (1~100위)
try:
    top_games, _ = collection(
        Collection.TOP_GROSSING,
        category='GAME',
        country='kr',
        lang='ko'
    )
except Exception as e:
    print(f"크롤링 에러 발생: {e}")
    top_games = []

# 6. 헤더 및 데이터 정리
headers = ["날짜", "순위", "앱 이름", "개발사(퍼블리셔)", "전일 대비"]
new_rows = [headers]

for rank, game in enumerate(top_games[:100], start=1):
    title = game['title']
    developer = game.get('developer', '알 수 없음')
    
    # 전일 대비 순위 계산 로직 ('앱 이름'으로 어제 시트와 비교)
    if title in previous_ranks:
        prev_rank = previous_ranks[title]
        diff = prev_rank - rank
        if diff > 0:
            change = f"▲ {diff}"
        elif diff < 0:
            change = f"▼ {abs(diff)}"
        else:
            change = "-"
    else:
        change = "NEW"
        
    new_rows.append([today_str, rank, title, developer, change])

# 7. 완성된 데이터를 오늘 시트에 한 번에 쓰기
if len(new_rows) > 1:
    today_sheet.update(values=new_rows, range_name="A1")
    print(f"{today_str} 기준 Top 100 매출 순위 업데이트 완료! (시트명: {today_str})")
else:
    print("업데이트할 데이터가 없습니다.")
